Caltech-IPAC/firefly
# Firefly 1104 file upload view panel reusability

#### [Firefly-1104](https://jira.ipac.caltech.edu/browse/FIREFLY-1104)  
**Updates** **(11/9)**: 
- Fixed code post PR comments
- moved FieldGroup over to FileUploadDropdown. 
- showing the upload panel as a dialog from the "Add MOC" dialog by calling showUploadDialog() with acceptMoc = true. FileAnalysis in FileUploadViewPanel checks if the file **isMoc** and if acceptMoc = true, if it's not moc, display a warning (although right now the warning is displayed even before you try and upload a file - need to fix this). 
- After loading MOC from the upload panel, the upload panel still stays in place (showUploadDialog's onSubmit would ideally call dispatchHideDialog - but I've commented it out for now since that was interfering with resultSuccess's request object being undefined for some reason - looking into this as well). 
- ImageDisplayOption's selectedImages stays empty the first time you try and select multiple images after a file upload (which is why you don't see the two radio button options for loading "All images in one window" or "One extension image per window"). If you load the images, though, and go back to the upload panel and select multiple images now, the options show up. I've tried debugging this, but it hasn't worked yet. 
- Because of the pending issues/bugs described above, I have still not created an actual PR, but please let me know if you have any comments on these new changes @robyww @loitly 

#### Testing
 - https://fireflydev.ipac.caltech.edu/firefly-1104-fileuploadviewpanel-reusability/firefly/
 - go to "Upload" and upload a FITS image. 
 - **_Fixed_**: Select 1 or more images (selecting and loading tables doesn't work - I am looking into this). 
 - From Image search, choose image type: View HiPS Images and target: m1. 
 - Try and add a MOC layer, and click on the upload button. This should open as a new dialog. Try and upload a moc fits file as well as a non-MOC fits file (you should see a warning even after you try to upload this). Working on fixing the warning you see before uploading. 

**Updates (11/11)**: 
- Test build is the same: https://fireflydev.ipac.caltech.edu/firefly-1104-fileuploadviewpanel-reusability/firefly/
- In FileUploadViewPanel, I was able to dispatch values to the store to be able to use them in getNextState. This has fixed the problem of user going back after loading images/tables to the upload panel and the upload panel not remembering the user's selected images/tables. Now this works. 
- Added a text warning in the upload panel itself for user trying to upload a non-MOC file when they get to the upload panel from "Add MOC Layer". Would a popup warning be better? (to test this, from a HiPs/MOC image, click on "Add MOC Layer", and then attempt to upload a non-MOC file. Also upload a MOC file to test it works as expected and doesn't throw a warning). 
- Fixed the issue with dispatchHideDialog not working in conjunction with resultSuccess (the request object was empty). Now after loading MOC, the upload panel should close and your MOC/table will be loaded in the background as expected. 
- The issue with ImageDisplayOption and TableDisplayOption options not showing up on the first render (but if you load and go back to upload panel, the options show up) was fixed by removing useStoreConnector and calling getSelectedRows (for ImageDisplayOption) and getFileFormat (for TableDisplayOption) directly. I'm not entirely sure why, would like to discuss it @robyww. 
- Other than that, I think this ticket is nearly done. Pending any review comments, I can go ahead and create a PR. 

**Updates (11/17)**:
- cleaned up more of the code based on comments, using useFieldGroupMetaState and useFieldGroupValue now, got rid of some of the useEffects and dispatchComponentStateChange completely. Removed the use of groupKey where I could (but it's still being used in getNextState and also in the FileAnalysis component) 
- I had to continue using useStoreConnector while calling getNextState (even with useFieldGroupMetaState in place), because without useStoreConnector, getNextState wasn't being called on re-renders. 
-  clearReport() still makes use of dispatchValueChange (I tried using setUploadMetaInfo here, it didn't really work, but I may have used it incorrectly). Let me know if this still needs to be looked at. 
- Note the lack of use of useStoreConnector in both ImageDisplayOption and TableDisplayOption (I have commented out the lines where we were using useStoreConnector). This is what I had mentioned when we last spoke, if I use useStoreConnector while calling the getSelectedRows function, the options are not displayed the first time a user makes multiple selections in the upload panel. But if they load images/tables and go back to the upload panel and make multiple selections, the options are now displayed. Removing useStoreConnector fixed this issue (I am not sure why). 
- Also note the use of `acceptMoc` as an option now. This starts all the way in HiPSImageSelect.jsx, where showUploadDialog (in FileUploadDropdown.jsx) is called with `acceptMoc` set to `true`. Eventually, the FileAnalysis component is passed down the acceptMoc parameter in FileUploadViewPanel, and it checks if acceptMoc is true and if the uploaded file is a MOC file (using isMOCFitsFromUploadAnalsysis). 

- **Testing**: https://fireflydev.ipac.caltech.edu/firefly-1104-fileuploadviewpanel-reusability/firefly/
   - Test uploading a MOC (search for a HiPs image first) and uploading moc and non-moc files from the upload panel dialog
   - trying to upload non-moc files should give you a warning in the upload panel 
   - try uploading fits files from the regular upload panel, make multiple selections of images/tables. Load them. Go back to the upload panel to make sure your selections are still there, edit them as needed and Load again. Make sure everything works as expected. 

**Updates (11/29)**:
- Lots of new changes. Changed the call to getNextState based on your comments here @robyww. Also small changes like getting rid of table ui id's, changing uppercase variable names of non-constants to lowercase. 
- Added keepState to FieldGroup's context. Passing keepState=false for the HiPS upload dialog. 
- removing tables in useEffect cleanup in FileUploadViewPanel 
- In FileUploadProcessor's sendImageRequest function, I commented out the line that appends the extension list to a file name (e.g.: 3emus-1-01_r99_fe55_test_1,3,4,6). Is it okay, though, that now windows with the different number of images (all loaded into one window) will have the same name in the layers popup? 
- added a popup warning when you try to load a non-moc fits file from a HiPs/MOC upload dialog (the only issue here is - there's also a warning in the upload dialog when the user uploads a non-moc fits file (this warning comes from the FileAnalysis component). But there's a bit of a lag for the warning to show up, ~1-3 seconds. I am trying to see if there's a better way to display this warning). To test this, upload a non-moc fits file from the HiPs/MOC upload dialog. And then attempt to load it. 
- The HiPs/MOC upload dialog stays up after the warning (if you attempt to load a non-MOC fits file). 
- got rid of the old showUpload function in HiPSImageSelect.jsx 

- **Testing (11/29)**: https://fireflydev.ipac.caltech.edu/firefly-1104-fileuploadviewpanel-reusability/firefly/
   - Make sure the regular upload panel works as expected. 
   - Upload a fits moc file, non-moc fits file, load multiple images/table, go back to the upload panel to make sure your previous selections show up. 
   - Then attempt to upload a non-MOC fits file from the HiPS/Moc upload dialog (should give a warning in the dialog itself). If you attempt to load this, you should get a new popup warning.
